core/src/main/java/kafka/log/remote/RemoteLogManager.java

```diff
@@ -937,7 +934,9 @@ private boolean isSegmentBreachedByRetentionSize(RemoteLogSegmentMetadata metada
                     }
                 }
                 if (shouldDeleteSegment) {
-                    logStartOffset = OptionalLong.of(metadata.endOffset() + 1);
+                    if (!logStartOffset.isPresent() || logStartOffset.get() < metadata.endOffset() + 1) {
+                        logStartOffset = Optional.of(metadata.endOffset() + 1);
```

Comment:
just curious. `logStartOffset` is used to update LSO, and it seems the method `maybeIncrementLogStartOffset` is monotonic already. So why the non-monotonic increase causes incorrect LSO?

https://github.com/apache/kafka/blob/5a5a2921462fedde664ec2a61b033240160b66f7/core/src/main/scala/kafka/log/UnifiedLog.scala#L997

Reply:
I've updated the title of the PR. It is not about non-monotonic LSO update instead there is an issue in computing the log-start-offset while deleting the remote log segments.